syntax = "proto3";
option cc_generic_services = true;
option java_generic_services = true;

import "common.proto";

package finance;

message CategoryInfo
{
   common.UniqueId id = 1;
   string name = 2;
}

message CategoryItems
{
   repeated CategoryInfo items = 1;
}

enum EventType
{
   SPENDING = 0;
   EARNING = 1;
}

message EventInfo
{
   common.UniqueId event_id = 1;
   common.UniqueId category_id = 2;
   string details = 3;
   uint32 amount_euro_cents = 4;
   int32 timestamp = 5;
   EventType type = 6;
}

message EventItems
{
   repeated EventInfo items = 1;
}

enum FinanceStatus
{
   CRITICAL = 0;
   OK = 1;
   GOOD = 2;
   GREAT = 3;
   EXCELLENT = 4;
}

message FinanceReport
{
   int32 balance = 1;
   FinanceStatus status = 2;
}

service FinanceService 
{
   rpc GetCategories (common.EmptyMessage) returns (CategoryItems);
   rpc DeleteCategory (common.UniqueId) returns (common.OperationResultMessage);
   rpc ReplaceCategory (CategoryInfo) returns (common.OperationResultMessage);
   rpc AddCategory (CategoryInfo) returns (common.UniqueId);

   rpc DeleteEvent (common.UniqueId) returns (common.OperationResultMessage);
   rpc ReplaceEvent (EventInfo) returns (common.OperationResultMessage);
   rpc AddEvent (EventInfo) returns (common.UniqueId);

   rpc QueryEvents (common.TimeRange) returns (EventItems);
   rpc GetReport (common.EmptyMessage) returns (FinanceReport);
}